### PR TITLE
Support empty tuples.

### PR DESCRIPTION
**Pull request by amosbird**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Now empty tuples can be stored, queried, inserted just like other columns. The idea is to use `Tuple(Nothing)`. This addresses #38125. This is an alternative implementation of https://github.com/ClickHouse/ClickHouse/pull/42445. This might be a solution to https://github.com/ClickHouse/ClickHouse/issues/53574.

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 182aec4f782ad404cd5f91793eb483d4fa86d7ef with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55021/182aec4f782ad404cd5f91793eb483d4fa86d7ef/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>✅ success</td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>⏳ pending</td></tr>
<tbody>
</table>

**Review by novikd — COMMENTED** [review]
In general LGTM, but I need to think more if it's okay. This change will allow to store `Nothing` so we need to consider all the consequences.

Actions to do:
1. Add a test with `ORDER BY ()`
2. Modify the comment in the `DataTypes/DataTypeNothing.h`

File: src/DataTypes/Serializations/SerializationNothing.h [hunk 1]
```diff
@@ -1,30 +1,20 @@
...
-        throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Serialization is not implemented for type Nothing");
-    }
 public:
-    void serializeBinary(const Field &, WriteBuffer &, const FormatSettings &) const override                       { throwNoSerialization(); }
-    void deserializeBinary(Field &, ReadBuffer &, const FormatSettings &) const override                            { throwNoSerialization(); }
-    void serializeBinary(const IColumn &, size_t, WriteBuffer &, const FormatSettings &) const override             { throwNoSerialization(); }
-    void deserializeBinary(IColumn &, ReadBuffer &, const FormatSettings &) const override                          { throwNoSerialization(); }
-    void serializeText(const IColumn &, size_t, WriteBuffer &, const FormatSettings &) const override { throwNoSerialization(); }
-    void deserializeText(IColumn &, ReadBuffer &, const FormatSettings &, bool) const override    { throwNoSerialization(); }
+    /// These methods are used to serde empty tuples: Tuple(Nothing).
```
[the remark this answers is not in the document]
Comment (amosbird) [reply src/DataTypes/Serializations/SerializationNothing.h:11]: It's not a typo. serde is a well-known abbreviations of serialization/deserialization :)

**Comment by alexey-milovidov** [comment]
Interesting... I think, from a fundamental perspective, an empty Tuple is not the same as a Tuple of a single Nothing element. What are the main difficulties in supporting pure empty tuples?

**Comment by amosbird** [comment]
> This change will allow to store Nothing so we need to consider all the consequences.

We can disable support for storing, same as we did for Nullable(Nothing) and Array(Nothing).

DB::Exception: Data type Nullable(Nothing) cannot be used in tables. (DATA_TYPE_CANNOT_BE_USED_IN_TABLES)
DB::Exception: Data type Array(Nothing) cannot be used in tables. (DATA_TYPE_CANNOT_BE_USED_IN_TABLES)

**Comment by amosbird** [comment]
> What are the main difficulties in supporting pure empty tuples?

There are several places that assume tuple is non-empty. It makes life easier to keep that invariant. It's also consistent with Nullable(Nothing) and Array(Nothing).

**Comment by novikd** [comment]
> What are the main difficulties in supporting pure empty tuples?

I totally agree with @amosbird. It requires the implementation of a special column type for `Tuple()` or some ad-hock solution like that. We must be able to send an empty tuple from server to client, so we need to be able to serialize/deserialize it.

> We can disable support for storing, same as we did for Nullable(Nothing) and Array(Nothing).

Yes, I'd prefer to make it consistent. It doesn't make much sense to store the value.

**Comment by amosbird** [comment]
> It doesn't make much sense to store the value.

Well it does make sense if we're using tuples and arrays to mimic JSON. In that case, empty tuple means empty object.

**Comment by novikd** [comment]
> Well it does make sense if we're using tuples and arrays to mimic JSON. In that case, empty tuple means empty object.

For that purpose, I'd prefer something like #53575.

**Comment by amosbird** [comment]
Heh, how is "dummy" better than nothing :)

**Comment by alexey-milovidov** [comment]
> It's also consistent with Nullable(Nothing) and Array(Nothing).

It is a different thing. Tuple of Nothing is still a tuple containing a single data type. For example, compare these:
```
Tuple()
Tuple(Nothing)
Tuple(Nothing, Nothing)
```

Only the empty Tuple type can have values (the empty tuple).
While Tuple(Nothing) cannot store any values because the Nothing data type cannot have values.

And compare these:
```
Tuple(Nullable(Nothing))
Tuple(Nullable(Nothing), Nullable(Nothing))
```

The first type can store values, and the only possible value is `(NULL)`.
The first type can store values, and the only possible value is `(NULL, NULL)`.

**Comment by amosbird** [comment]
> It is a different thing.

I agree. If we need a real empty tuple implementation, one main difficulty will be changing the layout of ColumnTuple. At least it should contain a size field.

On the other hand, although it's different, `Tuple(Nothing)` still looks good to me. It's a tuple that contains nothing, which is grammatically equivalent to an empty tuple :)

**Comment by alexey-milovidov** [comment]
Yes, it's ok to change the layout of ColumnTuple by introducing a size field.

But when you serialize/deserialize - there is no need for a size field (it will be difficult) - you can make it a special case with every empty tuple element serialized as a single zero byte.

**Pull request closed without merging**